From 1211d72bdbef659d5c580f8321760410958e0f0c Mon Sep 17 00:00:00 2001
From: Johannes Thumshirn <jthumshirn@suse.de>
Date: Tue, 3 May 2016 15:26:53 +0200
Subject: [PATCH] block: initialize hd_stuct's reference before assigning it
References: bsc#976705
Patch-Mainline: Submitted, 2016-05-3

Inititialize the hd_struct's perpcu reference before assigning the hd_struct
to the partition table list.

This fixes a race which could be triggered using a simple partition
creation/deletion loop with virtio-blk on aarch64.

Signed-off-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 block/partition-generic.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/block/partition-generic.c b/block/partition-generic.c
index 5d87019..aa5b83a 100644
--- a/block/partition-generic.c
+++ b/block/partition-generic.c
@@ -361,6 +361,10 @@ struct hd_struct *add_partition(struct gendisk *disk, int partno,
 			goto out_del;
 	}
 
+	err = -ENOMEM;
+	if (hd_ref_init(p))
+		goto out_free_info;
+
 	/* everything is up and running, commence */
 	rcu_assign_pointer(ptbl->part[partno], p);
 
@@ -368,8 +372,7 @@ struct hd_struct *add_partition(struct gendisk *disk, int partno,
 	if (!dev_get_uevent_suppress(ddev))
 		kobject_uevent(&pdev->kobj, KOBJ_ADD);
 
-	if (!hd_ref_init(p))
-		return p;
+	return p;
 
 out_free_info:
 	free_part_info(p);
-- 
1.8.5.2

